netty/src/main/java/io/grpc/transport/netty/AbstractWriteQueue.java

```diff
@@ -0,0 +1,180 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import com.google.common.base.Preconditions;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.Nullable;
+
+/**
+ * A queue of pending writes to a {@link Channel} that is flushed as a single unit.
+ */
+abstract class AbstractWriteQueue {
+
+  private static final int DEQUE_CHUNK_SIZE = 128;
+
+  /**
+   * {@link Runnable} used to schedule work onto the tail of the event loop.
+   */
+  private final Runnable later = new Runnable() {
+    @Override
+    public void run() {
+      channel.write(AbstractWriteQueue.this);
+    }
+  };
+
+  private final Channel channel;
+  private final BlockingQueue<QueuedCommand> queue;
+  private final AtomicBoolean scheduled = new AtomicBoolean();
+
+  /**
+   * ArrayDeque to copy queue into when flushing in event loop.
+   */
+  private final ArrayDeque<QueuedCommand> writeChunk =
+      new ArrayDeque<QueuedCommand>(DEQUE_CHUNK_SIZE);
+
+  public AbstractWriteQueue(Channel channel) {
+    this.channel = Preconditions.checkNotNull(channel, "channel");
+    queue = new LinkedBlockingQueue<QueuedCommand>();
+  }
+
+  /**
+   * Schedule a flush on the channel.
+   */
+  void scheduleFlush() {
+    if (scheduled.compareAndSet(false, true)) {
+      // Add the queue to the tail of the event loop so writes will be executed immediately
+      // inside the event loop. Note DO NOT do channel.write outside the event loop as
+      // it will not wake up immediately without a flush.
+      channel.eventLoop().execute(later);
+    }
+  }
+
+  /**
+   * Enqueue a write command on the channel.
+   *
+   * @param command a write to be executed on the channel.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, boolean flush) {
+    enqueue(command, null, flush);
+  }
+
+  /**
+   * Enqueue a write command with a completion listener.
+   *
+   * @param command a write to be executed on the channel.
+   * @param listener to be notified on the completion of the write.
+   * @param flush true if a flush of the write should be schedule, false if a later call to
+   *              enqueue will schedule the flush.
+   */
+  void enqueue(Object command, ChannelFutureListener listener, boolean flush) {
```

Comment:
I misread why we were passing in the listener instead of a promise. It doesn't actually seem like we are doing it for any particular reason. Am I missing something?